nomad/job_endpoint_hook_connect.go

```diff
@@ -570,6 +570,7 @@ func groupConnectValidate(g *structs.TaskGroup) error {
 
 func groupConnectUpstreamsValidate(g *structs.TaskGroup, services []*structs.Service) error {
 	listeners := make(map[string]string) // address -> service
+	sockets := make(map[string]string) // socket -> path
```

Comment:
We should be able to safely reuse the listeners map here (this would have been `// path -> service`, as well), which will save us some allocations.

```suggestion
	listeners := make(map[string]string) // address or path -> service
```

That'll let you reduce some of the branches below as well